File: backend/app/engine/layers/zones.py
"""
Layer 2: Zone Identifier.

Finds potential entry zones using Fibonacci retracement
and VWAP deviation bands.
"""

from dataclasses import dataclass

import numpy as np
import pandas as pd

from app.engine.indicators import calculate_fib_levels, compute_vwap
from app.engine.layers.trend import Trend, TrendResult


@dataclass
class EntryZone:
    zone_type: str
    upper: float
    lower: float
    strength: float
    levels: dict[float, float] | None = None

    def to_dict(self) -> dict:
        return {
            "zone_type": self.zone_type,
            "upper": self.upper,
            "lower": self.lower,
            "strength": self.strength,
        }


class ZoneIdentifier:
    """
    Layer 2: Finds potential entry zones using Fibonacci, VWAP bands,
    and support/resistance levels.
    """

    def find_zones(
        self, candles: pd.DataFrame, trend: TrendResult
    ) -> list[EntryZone]:
        zones: list[EntryZone] = []
        zones.extend(self._fibonacci_zones(candles, trend))
        zones.extend(self._vwap_deviation_zones(candles))
        return zones

    def _fibonacci_zones(
        self, candles: pd.DataFrame, trend: TrendResult
    ) -> list[EntryZone]:
        swing_high_idx = candles["high"].idxmax()
        swing_low_idx = candles["low"].idxmin()
        swing_high = float(candles["high"].loc[swing_high_idx])
        swing_low = float(candles["low"].loc[swing_low_idx])

        if swing_high == swing_low:
            return []

        if trend.direction == Trend.BULLISH:
            fib_levels = calculate_fib_levels(swing_low, swing_high)
        else:
            fib_levels = calculate_fib_levels(swing_high, swing_low)

        golden_upper = fib_levels[0.382]
        golden_lower = fib_levels[0.618]

        if golden_upper < golden_lower:
            golden_upper, golden_lower = golden_lower, golden_upper

        return [
            EntryZone(
                zone_type="fibonacci_golden",
                upper=float(golden_upper),
                lower=float(golden_lower),
                strength=0.7,
                levels=fib_levels,
            )
        ]

    def _vwap_deviation_zones(self, candles: pd.DataFrame) -> list[EntryZone]:
        vwap = compute_vwap(candles)
        std = candles["close"].rolling(20).std()

        if std.iloc[-1] is None or np.isnan(std.iloc[-1]):
            return []

        vwap_val = float(vwap.iloc[-1])
        std_val = float(std.iloc[-1])

        return [
            EntryZone(
                zone_type="vwap_1sigma",
                upper=vwap_val + std_val,
                lower=vwap_val - std_val,
                strength=0.5,
            ),
            EntryZone(
                zone_type="vwap_2sigma",
                upper=vwap_val + 2 * std_val,
                lower=vwap_val - 2 * std_val,
                strength=0.8,
            ),
        ]

File: supabase/functions/_shared/engine/zones.ts
/**
 * Layer 2: Zone Identifier.
 *
 * Finds potential entry zones using Fibonacci retracement
 * and VWAP deviation bands.
 */

import { calcFibLevels, calcVWAP, extractArrays, type OHLCV } from '../indicators.ts'
import { Trend, type TrendResult } from './trend.ts'

// ── Entry Zone ─────────────────────────────────────────────────
export interface EntryZone {
  zoneType: string
  upper: number
  lower: number
  strength: number
  levels?: Record<string, number> | null
}

export function entryZoneToDict(zone: EntryZone): Record<string, unknown> {
  return {
    zone_type: zone.zoneType,
    upper: zone.upper,
    lower: zone.lower,
    strength: zone.strength,
  }
}

// ── Zone Identifier ────────────────────────────────────────────
export class ZoneIdentifier {
  findZones(candles: OHLCV[], trend: TrendResult): EntryZone[] {
    const zones: EntryZone[] = []
    zones.push(...this._fibonacciZones(candles, trend))
    zones.push(...this._vwapDeviationZones(candles))
    return zones
  }

  private _fibonacciZones(candles: OHLCV[], trend: TrendResult): EntryZone[] {
    const { highs, lows } = extractArrays(candles)

    // Find swing high and swing low
    let swingHigh = -Infinity
    let swingLow = Infinity
    for (const h of highs) {
      if (h > swingHigh) swingHigh = h
    }
    for (const l of lows) {
      if (l < swingLow) swingLow = l
    }

    if (swingHigh === swingLow) {
      return []
    }

    let fibLevels: Record<string, number>
    if (trend.direction === Trend.BULLISH) {
      fibLevels = calcFibLevels(swingHigh, swingLow)
    } else {
      // calcFibLevels(high, low) — for bearish, swap so retracements go the other direction
      fibLevels = calcFibLevels(swingLow, swingHigh)
    }

    let goldenUpper = fibLevels['0.382']
    let goldenLower = fibLevels['0.618']

    if (goldenUpper < goldenLower) {
      ;[goldenUpper, goldenLower] = [goldenLower, goldenUpper]
    }

    return [
      {
        zoneType: 'fibonacci_golden',
        upper: goldenUpper,
        lower: goldenLower,
        strength: 0.7,
        levels: fibLevels,
      },
    ]
  }

  private _vwapDeviationZones(candles: OHLCV[]): EntryZone[] {
    const { closes } = extractArrays(candles)
    const vwap = calcVWAP(candles)

    // Rolling 20-bar standard deviation of close
    if (closes.length < 20) {
      return []
    }

    const last20 = closes.slice(-20)
    const mean = last20.reduce((s, v) => s + v, 0) / last20.length
    const variance = last20.reduce((s, v) => s + (v - mean) ** 2, 0) / last20.length
    const stdVal = Math.sqrt(variance)

    if (isNaN(stdVal) || stdVal === 0) {
      return []
    }

    const vwapVal = vwap[vwap.length - 1]

    return [
      {
        zoneType: 'vwap_1sigma',
        upper: vwapVal + stdVal,
        lower: vwapVal - stdVal,
        strength: 0.5,
      },
      {
        zoneType: 'vwap_2sigma',
        upper: vwapVal + 2 * stdVal,
        lower: vwapVal - 2 * stdVal,
        strength: 0.8,
      },
    ]
  }
}
